core/src/main/java/org/apache/iceberg/PartitionSpecParser.java

```diff
@@ -147,7 +156,12 @@ private static void buildFromJsonFields(PartitionSpec.Builder builder, JsonNode
       String transform = JsonUtil.getString(TRANSFORM, element);
       int sourceId = JsonUtil.getInt(SOURCE_ID, element);
 
-      builder.add(sourceId, name, transform);
+      // partition field ids are missing in old PartitionSpec, they always auto-increment from PARTITION_DATA_ID_START
+      if (!hasFieldId) {
```

Comment:
How about the forward compatibility? Is it possible that an old reader reads the new spec? then it still parses the new spec field id start from 1000?

Reply:
Instead of making these changes here, I think that this should be verified in `TableMetadata`. That would accomplish the same thing, but make it easier to check the table version.